Misc scenario edits (#936)
grafana/k6-docs: src/data/markdown/translated-guides/en/02 Using k6/14 Scenarios/02 Advanced Examples.md

excerpt: 'Advanced Examples using the k6 Scenario API - Using multiple scenarios, different environment variables and tags per scenario.'
---

You can use multiple scenarios in one script, and these scenarios can be run in sequence or in parallel.
Some ways that you can combine scenarios include the following:

> - Sequence workload start times

Reviewer: ```suggestion
- Have different start times to sequence workloads
```

- Add per-scenario tags and environment variables
- Make scenario-specific thresholds.

> - Run scenarios in different VU [lifecycle functions](/using-k6/test-lifecycle),

Author: ```suggestion
- Run scenarios in different VU [lifecycle functions](/using-k6/test-lifecycle).
```

> - Run scenarios in different VU [lifecycle functions](/using-k6/test-lifecycle).

Reviewer: I don't understand what you mean here, sorry :confused: You can't have different `setup()` or `teardown()` per scenario (see https://github.com/grafana/k6/issues/1638, though it will probably be solved by actually implementing https://github.com/grafana/k6/issues/1342)

Author: I'm trying to communicate what is done in this example, by the way.
Different scenarios run different VU logic.

```js
export function contacts() {
  http.get('https://test.k6.io/contacts.php', {
    tags: { my_custom_tag: 'contacts' },
  });
}

export function news() {
  http.get('https://test.k6.io/news.php', { tags: { my_custom_tag: 'news' } });
}
```

> - Run scenarios in different VU [lifecycle functions](/using-k6/test-lifecycle).

Author: ```suggestion
-  Export multiple scenarios as functions in [VU code](/using-k6/test-lifecycle)
```

Author: Scenario, more precise description

> - Export multiple scenarios as functions in [VU code](/using-k6/test-lifecycle).

Reviewer: ```suggestion
- Execute different exported functions in different scenarios, [not just the `default` function](https://k6.io/docs/using-k6/test-lifecycle/#the-vu-stage).
```

And https://k6.io/docs/using-k6/test-lifecycle/#the-vu-stage itself should probably be slightly expanded in another PR, to contain a mention (with a link here) about the same capability of scenarios.

And, this should probably be a separate issue and a bigger discussion, but "VU code" and "VU stage" in https://k6.io/docs/using-k6/test-lifecycle/ is probably more confusing than helpful :thinking: It is conflating the terms, "VU" has one meaning in every other context and another meaning in the "test lifecycle" document :thinking: 

It's probably more accurate to say "test/VU initialization context" and "test execution context" :thinking: Again, though, this is a bigger discussion out of scope for this PR.

Author: I worry this suggestion is a little to extract. IMO, the "killer feature" here is that different scenarios can execute different test logic. Do you agree? If so, how about

```diff
- Export multiple scenarios as functions in [VU code](/using-k6/test-lifecycle).
+ Use multiple scenarios to run different test logic, so that VUs don't run in only the `default` function](https://k6.io/docs/using-k6/test-lifecycle/#the-vu-stage).
```

Reviewer: :+1: yeah, this is better, though you can drop the `in` in of `run in only the default`, it should be `run only the default`

> Use multiple scenarios to run different test logic, so that VUs don't run only the `default` function](https://k6.io/docs/using-k6/test-lifecycle/).

Reviewer: ```suggestion
 - Use multiple scenarios to run different test logic, so that VUs don't run only the `default` function](https://k6.io/docs/using-k6/test-lifecycle/).
```

excerpt: 'Advanced Examples using the k6 Scenario API - Using multiple scenarios, different environment variables and tags per scenario.'
---

> Besides making it easier to model workloads, scenarios have a second benefit of adding separation of test logic.

Author: ```suggestion
Besides making it easier to model workloads, scenarios can also help you separate of test logic in one script.
```

> As you can split scenarios across in different VU [lifecycle functions](/using-k6/test-lifecycle),
> you can use scenarios to:
> - Sequence workloads
> - Add per-scenario tags and environment variables
> - Make scenario-specific thresholds.

Author: ```suggestion
Some ways that you can use scenarios include the following:
- Sequence workloads
- Add per-scenario tags and environment variables
- Make scenario-specific thresholds.
- Run scenarios in different VU [lifecycle functions](/using-k6/test-lifecycle),

```

excerpt: 'Advanced Examples using the k6 Scenario API - Using multiple scenarios, different environment variables and tags per scenario.'
---

You can use multiple scenarios in one script, and these scenarios can be run in sequence or in parallel.
So besides helping you model workloads, scenarios can also help you make more flexible and composable test logic.
Some ways that you can use scenarios include the following:

> - Sequence workloads

Reviewer: The problem with this (and somewhat with "flexible and composable test logic" above) is that it's not actually that easy to run scenarios "in sequence" :disappointed: we don't have a `startAfter` logic for scenarios (and we probably won't ever have it), just `startTime`... 

Related k6 issue: https://github.com/grafana/k6/issues/1342

> You can use multiple scenarios in one script, and these scenarios can be run in sequence or in parallel.
> So besides helping you model workloads, scenarios can also help you make more flexible and composable test logic.
> Some ways that you can use scenarios include the following:
> - Sequence workloads
> - Add per-scenario tags and environment variables
> - Make scenario-specific thresholds.

Author: ```suggestion
You can use multiple scenarios in one script, and these scenarios can be run in sequence or in parallel.
Some ways that you can combine scenarios include the following:
- Sequence workloads
- Add per-scenario tags and environment variables
- Make scenario-specific thresholds.
```

- Run scenarios in different VU [lifecycle functions](/using-k6/test-lifecycle),

> ## Sequence multiple scenarios

Reviewer: The problem with this (and with "Sequence workloads" and "flexible and composable test logic" above) is that it's not actually that easy to run scenarios "in sequence" :disappointed: We don't have a `startAfter` logic for scenarios (and we probably won't ever have it), just `startTime`... Which is not sequencing, it's a time offset

Related k6 issue: https://github.com/grafana/k6/issues/1342

Author: Oh, I thought I posted responses to this and the lines comment. I guess I deleted them when I did my own review :confused: 

With `maxDuration` can't you specify sequences? I see the limitations: you have to set "time thresholds" for the scenario, and you can't configure scenarios to immediately run one after the other, but you can sequence. Maybe we should just state this limitation (if I'm not mistaken with how it works).

Author: Ah, I see that not every scenario has a max duration.

Author: I tried to fix this with [18f5b76](https://github.com/grafana/k6-docs/pull/936/commits/18f5b767ce34218ff39bf40279e144659eb82dc7). I don't want to overstate or give incorrect information.

If I still am overstating sequencing, let's just delete those lines. But it does seem like the the arrival rate executors have set durations. So I don't understand why you couldn't set one executor to last 5 minutes and another to start after 5 minutes.

Reviewer: It is mostly fine now, I think, I just made a tiny change suggestion inline

> ## Sequence multiple scenarios

Author: ```suggestion
## Combine scenarios
```

> To sequence scenarios, you can use `startTime` property (in combination with the options specific to your scenario executor).

Author: ```suggestion
With the `startTime` property, you can configure your script to start some scenarios later than others. If you combine this with the executor's duration options, you can sequence your scenarios (this is easiest to do with executors with set durations, like the arrival-rate executors.).
```

> This configuration first executes a scenario where 50 VUs try to run as many iterations
> as possible for 30 seconds.
> It then runs the next scenario, which executes 100 iterations per VU for a maximum duration of 1 minute.

Reviewer: ```suggestion
This configuration first executes a scenario where 50 VUs try to run as many iterations as possible for 30 seconds. It then runs the next scenario, which executes 100 iterations per VU for a maximum duration of 1 minute.
```

Please don't insert extra newlines in the middle of paragraphs. It makes reading the code in different resolutions and making change suggestions much worse, see https://github.com/grafana/k6-docs/pull/229#discussion_r585350973

Reviewer: ![example](https://user-images.githubusercontent.com/3336773/207317195-5ed95524-8d93-4134-8a5d-32e2da956ed6.png)
e.g. this is how a diff in one of the lines below can look like, if I tile my windows ~60:40

Author: Can't believe I somehow deleted my entire screed about this holy war. In short, I disagree. For text I try to break lines on the sentence level, with splits on other punctuation if it gets too long.

First, for reasons of writer comfort. [Asciidoctor has a good list](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line). To expand a bit:
- In modal editors, it's much easier to yank sentences in, out, up and down if they are own each line.
- It provides visual cue about bloated paragraphs. More than 5 sentences is getting dense for our docs. It's easy to track, just count 5 lines.
- It provides visual cues about bloated sentences. 80 chars, the standard terminal space, makes up about 14 words in an English sentence. By coincidence, this is a good average to shoot for simple sentences (complex sentences, I.e. those that have dependant clauses, e.g."If you want to..., then you can...", can be longer).
- It avoids awful diffs from editor reflows.

About diffs, one sentence per line makes diffs much easier to look at, because the changes appear more closely together. Compare a commit from this PR with a commit I made on March 17. Each changes only a punctuation mark:
![image](https://user-images.githubusercontent.com/47385188/207360327-3cfd3171-5db4-4689-b942-1802464d05e3.png)

Sorry, I don't understand what you mean 60:40? Is that line length? I think 80 is a reasonable expectation. I do see that it's annoying to soft wrap. But I favor going farther the other way, with [semantic line breaks](https://sembr.org/). That way, the editors could look at every constituent of the sentence on the molecular level. However, I haven't reached this level of enlightment yet.

Reviewer: One sentence per line is mostly ok, but the screenshot I posted had a bunch of breaks mid-sentence. 

I still think the visual cues you'd get if your raw text was similar to how it will appear when the markdown is rendered are more relevant than any you'd get if you split the sentences apart, but it's not that big of a deal :man_shrugging:

</CodeGroup>

## Use different environment variables and tags per scenario.

> The previous example set tags on individual HTTP request metrics.

Reviewer: ```suggestion
The previous example sets tags on individual HTTP request metrics.
```

or "we set"?

</CodeGroup>

## Use different environment variables and tags per scenario.

> The previous example set tags on individual HTTP request metrics, but this
> can also be done per scenario, which would apply them to other
> [taggable](/using-k6/tags-and-groups#tags) objects as well.

Author: ```suggestion
The previous example set tags on individual HTTP request metrics.
But, you can also set tags per scenario, which applies them to other
[taggable](/using-k6/tags-and-groups#tags) objects as well.
```

</CodeGroup>

<Blockquote mod="note" title="">

> By default, a `scenario` tag, whose value is the scenario name, is
> applied to all metrics in each scenario.
> You can combine these tags with thresholds,
> or use them to simplify metric filtering in [result outputs](/get-started/results-output).

Author: ```suggestion
By default, k6 applies a `scenario` tag to all metrics in each scenario (the value is the scenario name.
You can combine these tags with thresholds, or use them to simplify results filtering.
```

</CodeGroup>

<Blockquote mod="note" title="">

> By default, a `scenario` tag with the name of the scenario as value is

Author: ```suggestion
By default, a `scenario` tag, whose value is the scenario name, is
```

applied to all metrics in each scenario.
You can combine these tags with thresholds,
or use them to simplify metric filtering in [result outputs](/get-started/results-output).

> You can disable scenario tags with the [`--system-tags` option](/using-k6/options#system-tags).

Author: ```suggestion
To disable scenario tags, use the [`--system-tags` option](/using-k6/options#system-tags).
```
